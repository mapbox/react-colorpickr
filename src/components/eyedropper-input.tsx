import React, { useRef, useEffect } from 'react';
import themeable from 'react-themeable';
import { autokey } from '../autokey';

interface Props {
  theme: { [id: string]: string };
  onChange: (value: string) => void;
  disabled?: boolean;
}

function EyedropperInput({ onChange, theme, disabled = false }: Props) {
  const action = useRef<HTMLButtonElement>();
  // @ts-expect-error EyeDropper object is still experimental and not fully supported
  const dropper = useRef(new EyeDropper());
  const abortController = useRef(new AbortController());
  const themer = autokey(themeable(theme));

  useEffect(() => {
    return () => abortController.current.abort();
  }, []);

  const activateDropper = async (): Promise<void> => {
    if (action.current) {
      action.current.classList.add('is-active');
    }

    try {
      const result = await dropper.current.open({
        signal: abortController.current.signal
      });
      onChange(result.sRGBHex);
      if (action.current) {
        action.current.classList.remove('is-active');
      }
    } catch (e) {
      return null;
    }
  };

  const buttonProps = {
    ...themer('eyeDropper'),
    'data-testid': 'eye-dropper',
    ref: action,
    type: 'button',
    disabled
  };

  if (!disabled) {
    buttonProps.onClick = activateDropper;
  }

  return (
    <button {...buttonProps}>
      <svg xmlns="http://www.w3.org/2000/svg" {...themer('eyeDropperIcon')}>
        <path
          fillRule="evenodd"
          clipRule="evenodd"
          d="M5.793 15.5a.5.5 0 0 0 .353-.146L11.75 9.75 12 10a.707.707 0 1 0 1-1l-.25-.25 2.836-2.836a2 2 0 0 0 0-2.828l-.672-.672a2 2 0 0 0-2.828 0L9.25 5.25 9 5a.707.707 0 0 0-1 1l.25.25-5.604 5.604a.5.5 0 0 0-.146.353V15a.5.5 0 0 0 .5.5h2.793Zm-.293-1L11 9 9 7l-5.5 5.5v2h2Z"
        />
      </svg>
    </button>
  );
}

export { EyedropperInput };
