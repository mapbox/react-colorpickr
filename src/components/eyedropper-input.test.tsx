import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { EyedropperInput } from './eyedropper-input';

describe('EyedropperInput', () => {
  const user = userEvent.setup();

  const props = {
    theme: {},
    onChange: jest.fn()
  };

  test('renders', () => {
    const { baseElement } = render(<EyedropperInput {...props} />);
    expect(baseElement).toMatchSnapshot();
  });

  test('disabled', async () => {
    render(<EyedropperInput {...props} disabled={true} />);
    await user.click(screen.getByTestId('eye-dropper'));
    await waitFor(() => {
      expect(props.onChange).not.toHaveBeenCalled();
    });
  });

  test('activates element', async () => {
    render(<EyedropperInput {...props} />);
    await user.click(screen.getByTestId('eye-dropper'));
    await waitFor(() => {
      expect(props.onChange).toHaveBeenCalledWith('rgba(255, 255, 255, 0)');
    });
  });
});
