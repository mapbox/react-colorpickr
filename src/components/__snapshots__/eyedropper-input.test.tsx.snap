// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EyedropperInput renders 1`] = `
<body>
  <div>
    <button
      data-testid="eye-dropper"
      type="button"
    >
      <svg
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5.793 15.5a.5.5 0 0 0 .353-.146L11.75 9.75 12 10a.707.707 0 1 0 1-1l-.25-.25 2.836-2.836a2 2 0 0 0 0-2.828l-.672-.672a2 2 0 0 0-2.828 0L9.25 5.25 9 5a.707.707 0 0 0-1 1l.25.25-5.604 5.604a.5.5 0 0 0-.146.353V15a.5.5 0 0 0 .5.5h2.793Zm-.293-1L11 9 9 7l-5.5 5.5v2h2Z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</body>
`;
